services/test_analytics/ta_finish_upload.py

```diff
@@ -0,0 +1,275 @@
+    EnrichedPull,
+    fetch_and_update_pull_request_information_from_commit,
+    get_repo_provider_service,
+)
+from services.seats import ShouldActivateSeat, determine_seat_activation
+from services.test_analytics.ta_metrics import (
+    read_failures_summary,
+    read_tests_totals_summary,
+)
+from services.test_analytics.ta_process_flakes import KEY_NAME
+from services.test_analytics.ta_timeseries import (
+    TestInstance,
+    get_flaky_tests_dict,
+    get_pr_comment_agg,
+    get_pr_comment_failures,
+)
+from services.test_results import (
+    ErrorPayload,
+    FinisherResult,
+    TACommentInDepthInfo,
+    TestResultsNotificationFailure,
+    TestResultsNotificationPayload,
+    TestResultsNotifier,
+    should_do_flaky_detection,
+)
+
+log = logging.getLogger(__name__)
+
+
+def get_upload_ids(commitid: str) -> dict[int, ReportSession]:
+    return {
+        upload.id: upload
+        for upload in ReportSession.objects.filter(
+            report__commit__commitid=commitid,
+            state__in=["processed", "finished", "flake_processed"],
+        )
+    }
+
+
+def get_upload_error(upload_ids: list[int]) -> ErrorPayload | None:
+    error = UploadError.objects.filter(report_session_id__in=upload_ids).first()
+    if error:
+        return ErrorPayload(
+            error_code=error.error_code,
+            error_message=error.error_params.get("error_message"),
+        )
+    return None
+
+
+def transform_failures(
+    uploads: dict[int, ReportSession], failures: list[TestInstance]
+) -> list[TestResultsNotificationFailure[bytes]]:
+    notif_failures = []
+    for failure in failures:
+        if failure["failure_message"] is not None:
+            failure["failure_message"] = shorten_file_paths(
+                failure["failure_message"]
+            ).replace("\r", "")
+
+        notif_failures.append(
+            TestResultsNotificationFailure(
+                display_name=failure["computed_name"],
+                failure_message=failure["failure_message"],
+                test_id=failure["test_id"],
+                envs=uploads[failure["upload_id"]].flag_names,
+                duration_seconds=failure["duration_seconds"] or 0,
+                build_url=uploads[failure["upload_id"]].build_url,
+            )
+        )
+    return notif_failures
+
+
+def queue_followup_tasks(
+    repo: Repository,
+    commit: Commit,
+    commit_yaml: UserYaml,
+    impl_type: Literal["new", "both"] = "both",
+):
+    if (
+        should_do_flaky_detection(repo, commit_yaml)
+        and commit.merged is True
+        and commit.branch == repo.branch
+    ):
+        redis_client = get_redis_connection()
+        redis_client.set(f"flake_uploads:{repo.repoid}", 0)
+        redis_client.lpush(KEY_NAME.format(repo.repoid), commit.commitid)
+
+        celery_app.send_task(
+            process_flakes_task_name,
+            kwargs={
+                "repo_id": repo.repoid,
+                "commit_id": commit.commitid,
```

Comment:
similar to the comment below, `commitid` and `commit_id` is confusing.
the first one is actually the `commit_sha`, and the `X_id` looks a lot like a primary key id, though it isn’t in this case.